analysis/data_utils.py

```diff
@@ -106,15 +106,17 @@ def filter_max_time(experiment_df, max_time):
     return experiment_df[experiment_df['time'] <= max_time]
 
 
-def add_bugs_covered_column(benchmark_df):
-    """Return a modified benchmark df in which the |crash_key| column is
-    replaced with a |bugs_covered| column."""
-    new_df = benchmark_df.drop(columns='crash_key').drop_duplicates()
-    new_df['bugs_covered'] = new_df.apply(lambda x: benchmark_df[
-        (benchmark_df['trial_id'] == x['trial_id']) &
-        (benchmark_df['fuzzer'] == x['fuzzer']) &
-        (benchmark_df['time'] <= x['time'])]['crash_key'].nunique(),
-                                          axis=1)
+def add_bugs_covered_column(experiment_df):
+    """Return a modified experiment df in which adds a |bugs_covered| column,
+    a cumulative count of bugs covered over time."""
+    grouping1 = ['fuzzer', 'benchmark', 'trial_id', 'crash_key']
+    grouping2 = ['fuzzer', 'benchmark', 'trial_id']
+    grouping3 = ['fuzzer', 'benchmark', 'trial_id', 'time']
+    df = experiment_df.sort_values(grouping3)
+    df['firsts'] = ~df.duplicated(subset=grouping1) & ~df.crash_key.isna()
+    df['bugs_cumsum'] = df.groupby(grouping2)['firsts'].transform('cumsum')
+    df['bugs_covered'] = df.groupby(grouping3)['bugs_cumsum'].transform('max')
```

Comment:
Wondering why we need to `max` and why can't we just `cumsum` the `firsts`:
```
df['bugs_covered'] = df.groupby(grouping3)['firsts'].transform('cumsum')
```

Reply:
`cumsum` gives the current count, incrementing for every first over the current group.  I couldn't figure out another way to cumulatively sum, but just report the max for each time period.  It seemed less error prone to have `bugs_covered` reported the same way `edges_covered` does in the dataframe so that they can be used interchangeably.